### DISCUSSION TITLE
[General] Delays in Transaction Confirmation: Reevaluating Time Measurement and Considerations for Improved User Experience

### DISCUSSION THREAD

**Discussion by gauravl-tevaeralabs** [zkSync-Community-Hub/zksync-developers#280] [General]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Discussing the delay in transaction confirmation attributed to the time required by zkSync to furnish the transaction receipt, which adversely affects the user experience in dApps. This delay has been consistently observed over the past month, extending the time required for the completion of even basic transactions to 10-15 seconds. It's worth noting that similar delays may be encountered across other dApps within the zkSync ecosystem.

### What is the specific issue or error you're encountering?

We calculate the time taken for the transaction and display it in the success alert. Below is the code snippet illustrating how we calculate this time and a snapshot of where we present it, impacting the user experience. Previously, it took less than a second; however, currently, it's taking 10-15 seconds, and this delay is reproducible in other dApps within the ecosystem.

Code snippet:
```
const contract = await new CitizenId({ web3Provider: blockchainProvider.current, network: getNetwork() });
const receipt = await contract.mintCitizenID(); // this might be faster
const initiateTime = Date.now();
const data = await receipt.wait(); // this is taking more time now (~10-15 sec)
const confirmationDate = Date.now();
```

![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/122139139/c8786b0d-e327-4990-b743-7ba5264043d5)

### Can you share the error messages or logs you're receiving, if any?

NA

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

NA

### Have you tried to isolate the problem, and if so, what were the results?

It appears that the generation of receipts has started to take more time.

### What steps have you already taken to try to resolve the issue?

So far, no action has been taken. I am hopeful that either team will resolve it if possible, or I am seeking suggestions on how to calculate the time taken by the transaction, as I've observed issues with the current mentioned method.

**Comment by MexicanAce** [comment]
Hi @gauravl-tevaeralabs :wave:!

Could you please post your code behind `mintCitizenID()` and `CitizenId`?

I've attempted to replicate this locally, and using `zksync-ethers`, I've been able to mint dozens of ERC721 tokens with an average `.wait()` of ~1.5 seconds. I've seen wait times of ~5 seconds with `ethers` directly, but that still doesn't line up with what you've mentioned.

Can you please also post the `package.json` for your code?

**Reply by gauravl-tevaeralabs** [reply]
Hi @MexicanAce,

Thank you for the prompt reply. The Solidity code behind mintCitizenID() is not complex; it simply involves erc721.mint() on the contract side. I'm currently using "zksync-web3@0.14.3" instead of "zksync-ethers". Although I'm aware of the need to upgrade to zksync-ethers, it shouldn't be the root cause of the performance issue.

As previously mentioned, this issue arose last month, and initially, the transaction time was less than 1 second. However, it has since slowed down. Additionally, it's important to note that this issue is not unique to our code; it's reproducible in many major projects within the zksync ecosystem. Some notable examples include mute.io, SyncSwap, Element Marketplace, etc. You can verify this by attempting a simple transaction on any of the mentioned dApps and observing the same issue.

Thanks again for your assistance.

**Reply by uF4No** [reply]
I just did some swaps on SyncSwap and Mute, all under 2secs
![Screenshot 2024-01-18 at 10 15 09](https://github.com/zkSync-Community-Hub/zksync-developers/assets/1957117/75bff4a0-ecea-49ea-8a0b-117e67c1d19f)

If you see similar issues in multiple apps, it could be something in your end. Which wallet are you using? Are you using the default RPC or one from one the the RPC providers?

**Reply by MexicanAce** [reply]
Marking the above response as the answer, if this continues as a problem that we can replicate, please let us know and we'll continue the investigation. Thank you!

**Reply by gauravl-tevaeralabs** [reply]
Hello @MexicanAce, I hope this message finds you well. I wanted to kindly inform you that I've decided to unmark the previous comment as an answer. We're currently encountering a delay issue across the zksync ecosystem. I mentioned earlier that this isn't a consistent occurrence. When we conduct multiple transactions, the duration can vary from 1 second to 15 seconds.

As @uF4No mentioned, despite receiving transaction receipts within 1-2 seconds in some cases, there are instances where the duration extends up to 15 seconds. In the worst-case scenario, it may take up to 20 seconds.

I have attached transactions conducted by different team members from various regions and on different websites for your reference. Thank you for your understanding and assistance in addressing this matter.

And here's the transaction hash for all the attached screenshots:
https://era.zksync.network/tx/0xf029b8c27951cb2c2401490733afe172ef11f803d86cad47f379d5f50d788a78
https://explorer.zksync.io/tx/0xe2f4ebb507ce4fd4beca94c0b5e959f90ebe318cac3a26dc745df6ba85b00388
https://explorer.zksync.io/tx/0xc6615e6d46f63989fd9f7e7fe8cb39cd074ac24e881aab411b263b3b1b7961a2
https://explorer.zksync.io/tx/0xf1dea887d2708d2e6cae5685902219f86bf530f80501168438e7615be45ae60e
https://explorer.zksync.io/tx/0xc79970ed10b02f2057f518185a5ff0b3ddd5e33f27384fe1e3fa9a6e1b1af9aa

![photo_2024-02-09_16-27-52](https://github.com/zkSync-Community-Hub/zksync-developers/assets/122139139/1a211192-03e9-4335-909a-da219a3c0d51)
![tg_image_1715966810](https://github.com/zkSync-Community-Hub/zksync-developers/assets/122139139/0b7a3cd7-a398-4ead-b2a1-5f8091bed8a8)
![tg_image_3936710197](https://github.com/zkSync-Community-Hub/zksync-developers/assets/122139139/b87bb6bf-e817-477d-9b8f-e64cead8e762)
![tg_image_163572392](https://github.com/zkSync-Community-Hub/zksync-developers/assets/122139139/6268578b-c7a6-4122-9adc-993955eff869)
![tg_image_3999548207](https://github.com/zkSync-Community-Hub/zksync-developers/assets/122139139/a5ada65d-a851-4c14-901b-b6e3d0a79032)

Note: All transactions were done using Metamask wallet with the latest version.

**Reply by RomanBrodetski** [reply]
Hello, thanks for the additional report!

First of all,
>  As previously mentioned, this issue arose last month, and initially, the transaction time was less than 1 second. However, it has since slowed down.

In December we changed the API behaviour to be more consistent with Ethereum - before that we were returning transaction recipes even if they are not yet included in a miniblock (albeit with a null block_hash field). Many clients treated this as transaction being confirmed. I believe this change resulted in you observing higher times.

> https://era.zksync.network/tx/0xf029b8c27951cb2c2401490733afe172ef11f803d86cad47f379d5f50d788a78
> https://explorer.zksync.io/tx/0xe2f4ebb507ce4fd4beca94c0b5e959f90ebe318cac3a26dc745df6ba85b00388
> https://explorer.zksync.io/tx/0xc6615e6d46f63989fd9f7e7fe8cb39cd074ac24e881aab411b263b3b1b7961a2
> https://explorer.zksync.io/tx/0xf1dea887d2708d2e6cae5685902219f86bf530f80501168438e7615be45ae60e
> https://explorer.zksync.io/tx/0xc79970ed10b02f2057f518185a5ff0b3ddd5e33f27384fe1e3fa9a6e1b1af9aa

I checked these transactions in the DB to compare the time it was received with the time it was included in block. As you can see, the times are very low.
![Screenshot 2024-02-09 at 15 29 59](https://github.com/zkSync-Community-Hub/zksync-developers/assets/1053184/9c7e6a1a-8994-489d-8a47-224154b1b93e)

There are also some API requests that `await receipt.wait()` does, but they are - normally - very fast. 

@MexicanAce also ran some tests to see how much time a simple mint takes on our side - the results are very different from what we see on your screenshot.
![Screenshot 2024-02-10 at 08 23 02](https://github.com/zkSync-Community-Hub/zksync-developers/assets/1053184/5f0ea656-ddf2-4e6b-867a-fb52a5a2a45a)

To debug further we need to figure out what exactly takes that much time. I'm not an expert in the frontend code, but let's try to measure the time it takes from your side to do various api calls. Instead of just `wait()` let's see how much time `estiamte_gas`, `get_tx_receipt` take on your side. Can you please do such an experiment and paste results? Thanks!

**Reply by gauravl-tevaeralabs** [reply]
Thanks for the insights and detailed explanation, @RomanBrodetski. Please let us know if anything is needed from our side.

**Reply by gauravl-tevaeralabs** [reply]
I'm uncertain, but is it possible that the issue stems from dApps still using _zksync-web3_ instead of the latest _zksync-ethers_, given that **wait()** is a function within the ethers.js library? We are presently in the process of transitioning from _zksync-web3_ to _zksync-ethers_, and we will inform you once we have confirmed whether this transition resolves the issue for our dApp.

**Reply by MexicanAce** [reply]
In a [recent announcement](https://github.com/zkSync-Community-Hub/zksync-developers/discussions/233), `zksync-web3` has been migrated to `zksync-ethers@5`.

As per your previous post (https://github.com/zkSync-Community-Hub/zksync-developers/discussions/322), you should migrate to `zksync-ethers@5` and make the appropriate adjustments.

**Reply by RomanBrodetski** [reply]
@gauravl-tevaeralabs any update on this? Are you still experiencing delays?

**Comment by edevaldonunez** [comment]
The delay in transaction confirmation on zkSync may be attributed to increased network activity or congestion. It's essential to consider optimizing the way transaction times are calculated and presented to users in dApps to mitigate the impact of these delays on user experience. One approach could be to implement client-side estimations based on the gas price and network conditions to provide more responsive feedback to users while waiting for transaction confirmation.

**Comment by jodychvs** [comment]
The delay in transaction confirmation on zkSync, lasting 10-15 seconds, adversely affects user experience in dApps. It's reproducible across the ecosystem. Current time calculation methods, reliant on transaction receipts, exacerbate the issue. Seeking resolution or alternative time calculation methods. #zkSync #UserExperience

**Comment by sonhoranemichelle** [comment]
The delay in transaction confirmation within zkSync dApps has been noted, affecting user experience. Consider optimizing the method for calculating transaction time or awaiting resolution from zkSync team.
